[docs]: add core package documentation (#18)
soramitsu/vue-kakuyaku: packages/core/README.md

- Focus on the best TypeScript support
- Minimal opt-in API
- Suitable for both "fetching data" and "doing side effect" cases
- Utilities:
  - [Retry-on-error](#retry-on-error-with-useerrorretry)
  - [Stale-if-error](#stale-if-error-with-usestalestate) state
  - [Shorthand watchers](#watcher-shorthands-for-the-state) for task results

>   - [Non-parametrised](#set-up-an-async-action-in-separate-scope-with-usedeferredscope) and [reactively parametrised](#reactively-parametrised-scope-with-useparamscope) scopes to model async operations stronger

Reviewer: ```suggestion
  - [Non-parametrised](#set-up-an-async-action-in-separate-scope-with-usedeferredscope) and [reactively parametrised](#reactively-parametrised-scope-with-useparamscope) scopes to model async operations in a more modular way
```

## Docs

### Atomic Promise State

The core primitive type of the library is `PromiseStateAtomic<T>`. It describes all possible states of a Promise in a given time in **type-safe** and **exclusive** manner.

`PromiseStateAtomic<T>` is a set of mutually exclusive invariants:

```ts
type PromiseStateAtomic<T> =
  // uninit
  | { rejected: null; fulfilled: null; pending: false }
  // pending
  | { rejected: null; fulfilled: null; pending: true }
  // fulfilled
  | { rejected: null; fulfilled: { value: T }; pending: false }
  // rejected
  | { rejected: { reason: unknown }; fulfilled: null; pending: false }
```

It is **type-safe** (relatively to other libraries which work with promises), because:

> - Empty `fulfilled` and `rejected` states are deterministically distinguished from their existence, even if the promise resolves/rejects with a nullable value.

Author: Hmph, maybe:

> You don't need to guess if `state.fulfilled === null` means that the promise is not fulfilled yet, or it is resolved with `null` value. `state.fulfilled` is `null | { some: T }`, so if there is *some* `state.fulfilled`, then the promise is fulfilled for sure. Same for `state.rejected`.

> - Empty `fulfilled` and `rejected` states are deterministically distinguished from their existence, even if the promise resolves/rejects with a nullable value.

Reviewer: ```suggestion
- Empty `fulfilled` and `rejected` states are easily distinguished. You don't need to guess if `state.fulfilled === null` means that the promise is not fulfilled yet or if it is resolved with `null` value. `state.fulfilled` is `null | { some: T }`, so if there is *some* `state.fulfilled`, then the promise is fulfilled. Same for `state.rejected`.
```

- `rejected.reason` is an `unknown`, not an `Error` or unsafe `any`, which is how JavaScript works - literally everything could be `throw`n.

It is **exclusive**, because with it TypeScript might know that the promise is either pending, rejected, fulfilled or empty, and not a mix of them. Thus, TypeScript narrows types on assertions:

```ts
declare const state: PromiseStateAtomic<string>

if (state.pending) {
  // no error - TypeScript narrows the types of
  // `rejected` and `fulfilled`
  const a: null = state.rejected || state.fulfilled
}
```

Having `PromiseStateAtomic<T>`, we can proceed to the utilities built around it.

### Basics with `usePromise()`

This composable gives a basic reactive model over any promise you put into it:

- Reactive `state` which is `PromiseStateAtomic<T>`
- Method to `set(promise: Promise<T>)` so its state will be reflected in `state`

> - Method to `clear()` the composable to the initial empty state.

Author: I changed this part and moved it forward, after the code section.

}
```

In this example, there is no need for `fulfilled` to be `null | { value: { bar: 'baz' } }` in order to distinguish empty fulfilled state from existing one - it would be enought to be just `null | { bar: 'baz' }`.

For such a case, the state could be reactively flattened when there is no need for nested `fulfilled.value` and `rejected.reason` fields:

```ts
const flattenedState = flattenState(state)

if (flattenedState.fulfilled) {
  const baz = state.fulfilled.bar
}
```

`mode` argument could be passed to control which fields are flattened:

```ts
// `all`, `fulfilled` (default) and `rejected` are accepted
flattenState(state, 'all')
```

### Retry on error with `useErrorRetry()`

This composable watches for the promise's state and, if it is rejected, invokes the callback (assuming that it will trigger the state to refresh) for a given amount of times with a given interval.

```ts
const { state, run } = useTask(
  async () => {
    if (Math.random() > 0.5) throw new Error('bad luck')
  },
  { immediate: true },
)

useErrorRetry(state, () => run(), {
  // default - 5
  count: 10,
  // default - 5000
  interval: 300,
})
```

### Stale if error with `useStaleState()`

> This composable is a tiny and lightweight implementation of [stale-while-revalidate](https://web.dev/stale-while-revalidate/) pattern. In short, it is about caching and using successful result while they are revalidating, even with failures.

Author: Oh, I meant "it is revalidating", i.e. the state of async task.

> This composable is a tiny and lightweight implementation of [stale-while-revalidate](https://web.dev/stale-while-revalidate/) pattern. In short, it is about caching and using successful result while they are revalidating, even with failures.

Reviewer: so like this?
```suggestion
This composable is a tiny and lightweight implementation of [stale-while-revalidate](https://web.dev/stale-while-revalidate/) pattern. In short, it is about caching and using successful result while the state of async task is revalidating, even with failures.
```

```ts
interface PromiseStaleState<T> {
  fulfilled: null | { value: T }
  rejected: null | { reason: unknown }
  pending: boolean
  fresh: boolean
}
```

When the _atomic_ state becomes _fulfilled_, the _stale_ state updates its `fulfilled` too, removing the last rejection reason if there were some. When the _atomic_ state becomes _rejected_, the _stale_ state only updates rejection reason, without touching the last `fulfilled` value.

`PromiseStaleState<T>` type is not [_exclusive_](#atomic-promise-state) - the stale state might have a fulfilled value, a rejection reason and be pending **at the same time**. You might think of it as a `PromiseStateAtomic<T>` with memory about its previous executions.

This utility might be useful in simple scenarios. However, it cannot be compared with libraries such as [`Kong/swrv`](https://github.com/Kong/swrv) or [`ConsoleTVs/vswr`](https://github.com/ConsoleTVs/vswr) - they implement SWR pattern in a much more comprehensive way. Anyway, they have their own drawbacks, thus `@vue-kakuyaku/swr` is planned to be a competitive solution.

### Set up an async action in separate scope with `useDeferredScope()`

> When it comes to modeling async actions within a component (or any other reactive scope), it might become a mess if action's lifetime is not the same as the component's one. It might be initialised on some event, or discarded on another. You might need to set up reactive stuff around this action (such as timers, retrying, showing notifications etc) during the **component's** setup stage.

Author: "logic" is better, imo.

)

function retryLogin() {
  loginScope.value?.expose.retry()
}
```

> We use `expose` keyword, trying to follow [existing Vue semantics](https://vuejs.org/api/options-state.html#expose) around this keyword.

This utility is a bit low-level and not very useful for a very common case - when you need to set up an async action, based on reactive parameters. The next section gives a solution for it.

### Reactively parametrised scope with `useParamScope()`

Such a common scenario: use have a reactive `userId` and you need to fetch a user data, according to the ID. Additionally, you might need to set up some reactive logic around it.

Here is how it might look with `useParamScope()`:

```ts
const userId = ref(1)

const scope = useParamScope(
  // reactive key - a ref or a getter
  userId,

  // setup function, that accepts the resolved key for the scope
  (staticUserId) => {
    const { state, run } = useTask(
      () => fetch(`/users/${staticUserId}`),
      { immediate: true },
    )

    useErrorRetry(state, run)
    const staleState = useStaleState(state)

    return staleState
  },
)

const userData = computed(() => scope.value.expose.fulfilled?.value)
```

> Whenever the reactive (and **primitive**, i.e. `number | string | symbol | boolean`, such as [Vue's `key` attr](https://vuejs.org/api/built-in-special-attributes.html#key) (except `boolean`)) key is changed, the existing scope is disposed, and the new one is set up.

Reviewer: ```suggestion
Whenever a reactive key is changed, the existing scope is disposed, and the new one is set up. The same applies for **primitive** keys: `number | string | symbol | boolean`.
```

not sure what this part means:
```
such as [Vue's `key` attr](https://vuejs.org/api/built-in-special-attributes.html#key) (except `boolean`))
```

Author: "The same applies..." doesn't seem correct.

I meant that the reactive *key* should be *primitive* anyway.

As for Vue: https://vuejs.org/guide/essentials/list.html#maintaining-state-with-key

> The `key` binding expects primitive values - i.e. strings and numbers. Do not use objects as v-for keys.

```html
<div v-for="i in 5" :key="i"">{{ i }}</div>
```

Vue developers know good about the `key` special attribute that Vue uses to distinguish elements from each other. Only primitive value should be bound to it, and the same rule applies for our keys.

Reviewer: ```suggestion
Whenever a reactive key is changed, the existing scope is disposed, and the new one is set up. The reactive keys should be **primitive** keys: `number | string | symbol | boolean`.
```

## Docs

### Atomic Promise State

The core primitive type of the library is `PromiseStateAtomic<T>`. It describes all possible states of a Promise at a given time in **type-safe** and **exclusive** manner.

`PromiseStateAtomic<T>` is a set of mutually exclusive invariants:

```ts
type PromiseStateAtomic<T> =

>   // uninit

Reviewer: ```suggestion
  // empty (uninitialised)
```

  | { rejected: { reason: unknown }; fulfilled: null; pending: false }
```

Compared to other libraries that work with promises, `PromiseStateAtomic<T>` is **type-safe** for the following reasons:

- Empty `fulfilled` and `rejected` states are easily distinguished. You don't need to guess if `state.fulfilled === null` means that the promise is not fulfilled yet or if it is resolved with `null` value. `state.fulfilled` is `null | { some: T }`, so if there is *some* `state.fulfilled`, then the promise is fulfilled. Same for `state.rejected`.
- `rejected.reason` is an `unknown`, not an `Error` or unsafe `any`, which is how JavaScript works: anything could be `throw`n.

`PromiseStateAtomic<T>` is **exclusive** because at any given moment promise is either pending, rejected, fulfilled or empty, and not a mix of them. Thus, TypeScript narrows types on assertions:

```ts
declare const state: PromiseStateAtomic<string>

if (state.pending) {
  // No error: TypeScript narrows the types of
  // `rejected` and `fulfilled`
  const a: null = state.rejected || state.fulfilled
}
```

Let's proceed to the utilities built around `PromiseStateAtomic<T>`.

### Basics with `usePromise()`

> This composable gives a basic reactive model over any promise you put into it:

Reviewer: ```suggestion
This composable returns a basic reactive model over a promise you put into it:
```

```ts
import { computed } from 'vue'

async function getString() {
  return '42'
}

const { state, set, clear } = usePromise<string>()

// `state` is a reactive object
// initially it is an empty state
const isPending = computed(() => state.pending)

// passing a `Promise<string>`
set(getString())

// forget the currently tracked promise if there is any
clear()
```

`usePromise<T>()` composable returns:

> - Reactive **`state`** which is `PromiseStateAtomic<T>`
> - Method to **set** a promise (`set(promise: Promise<T>)`) so its state will be reflected in `state`
> - Method to **clear** (`clear()`) the composable to the initial empty state.

Reviewer: ```suggestion
- Reactive **`state`** (`PromiseStateAtomic<T>`)
- The method to **set** a promise (`set(promise: Promise<T>)`) so its state is reflected in `state`
- The method to **clear** (`clear()`) the composable to the initial empty state
```

> **Note**: if a new promise is `set` while the previous one is pending, the result of the previous is ignored.

Reviewer: ```suggestion
**Note**: if a new promise is `set` while the previous one is pending, the result of the previous promise is ignored.
```
